### PR TITLE
Install torsten

### PR DESCRIPTION
**Pull request by billgillespie**
This pull request implements an `install_torsten` function. The function is similar to `cmdstanr::install_cmdstan` but with fewer options/arguments. I also added a test script that was adapted from one in cmdstanr. The current version of `install_torsten` does not automate the use of MPI. MPI is supported via the `cpp_options` argument.

### CODE REVIEW

File: R/install_torsten.R [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+#' @description The `install_torsten()` function attempts to download and
+#'   install the latest release of [Torsten](https://github.com/metrumresearchgroup/Torsten/releases/latest).
+#'   Installing a previous release or a new release candidate is also possible
+#'   by specifying the `version` or `release_url` argument.
+#'   See the first few sections of the CmdStan
+#'   [installation guide](https://mc-stan.org/docs/cmdstan-guide/cmdstan-installation.html)
+#'   for details on the C++ toolchain required for installing CmdStan.
+#'
+#'
+#' @export
```
Comment (kyleam) [inline R/install_torsten.R:12]: This docstring needs to be above `install_torsten` so that it's exported. As it is now, it exports `latest_torsten_release`.

As a related but more general note, the documentation needs to be regenerated and the `NAMESPACE` file updated. This can be done by calling `devtools::document()`.

File: R/install_torsten.R [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+#' @param cpp_options (list) Any makefile flags/variables to be written to
+#'   the `make/local` file. For example, `list("CXX" = "clang++")` will force
+#'   the use of clang for compilation.
+#' @param check_toolchain (logical) Should `install_torsten()` attempt to check
+#'   that the required toolchain is installed and properly configured. The
+#'   default is `TRUE`.
+#'
+#' @examples
+#' \dontrun{
+#' # install_torsten(cores = 4)
```
Comment (kyleam) [inline R/install_torsten.R:41]: Hmm, this example is code is marked with `dontrun`, so I think it's fine to leave `install_torsten` uncommented here. I see that `cmdstanr::install_cmdstan()` does similar commenting in its docstring, so perhaps there's a reason.

Anyway, fine to leave as is.

File: R/install_torsten.R [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+#'   PRECOMPILED_HEADERS = TRUE
+#' )
+#' # install_torsten(cores = 4, cpp_options = cpp_options)
+#' }
+#'
+
+#'## install_torsten version without unexported dependencies
+
+latest_torsten_release <- function() {
+  require(jsonlite)
```
Comment (kyleam) [inline R/install_torsten.R:55]: There are a few spots that use `require` (two for jsonlite and one for cmdstanr). Package code shouldn't call `require` because doing so modifies the user's search path.

For the jsonlite ones, all the function calls are already prefixed, so the two `require(jsonlite)` lines can just be deleted.

For the cmdstanr one, in addition to dropping the `require(cmdstanr)` line, two cmdstanr functions need to be prefixed:

-   [set\_cmdstan\_path call](https://github.com/metrumresearchgroup/bbr.bayes/blob/92653d994f30e7ab14e032069283bc5ff355a43b/R/install_torsten.R#L191)
-   [cmdstan\_model call](https://github.com/metrumresearchgroup/bbr.bayes/blob/92653d994f30e7ab14e032069283bc5ff355a43b/R/install_torsten.R#L193-L194)

File: R/install_torsten.R [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
```
Comment (kyleam) [inline R/install_torsten.R:56]: Ideally `dest_file` would be removed after its no longer needed. Could use [withr::local\_tempfile](https://withr.r-lib.org/reference/with_tempfile.html) helper here.

File: R/install_torsten.R [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  sapply(release,
+         function(x){
+           x$tag_name
+         })
+}
+
+install_torsten <- function(dir = NULL,
+                            cores = getOption("mc.cores", 2),
+                            quiet = FALSE,
+                            timeout = 1200,
```
Comment (kyleam) [inline R/install_torsten.R:88]: A description for `timeout` is missing from the documentation.

Side remark: a lot of the arguments are shared with `cmdstanr::install_cmdstan`, so I was going to mention that you could use `@inheritParams cmdstanr::install_cmdstan` to avoid needing to repeat a lot of the descriptions. However, I see that most of the descriptions won't quite work as is (e.g., "Torsten" is used instead of "cmdstanr").

It's not all or none, so you could still use `@inheritParams` for the ones that are identical (looks like `cpp_options` and `cores`), but it's fine as is.

File: R/install_torsten.R [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                            version = NULL,
+                            release_url = NULL,
+                            cpp_options = list(),
+                            check_toolchain = TRUE
+) {
+
+  ## install_torsten is based on cmdstanr::install_cmdstan. Differences include
+  ## * install_torsten always overwrites pre-existing Torsten installation at the
+  ##.  specified cmdstan_path.
```
Comment (kyleam) [inline R/install_torsten.R:97]: By default, I don't think `install_torsten` should overwrite an existing directory. I'm in favor of adding an `overwrite` argument and setting it to `FALSE` be default.

File: R/install_torsten.R [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+  ## * install_torsten does not try to automatically append to cpp_options for
+  ##.  M1 Macs. Is this needed?
+  ## * install_torsten does not provide support for WSL. Should it?
```
Comment (kyleam) [inline R/install_torsten.R:100]: This is a convenience helper, so I think it's fine for it to target the main case we're interested in and punt on supporting these. However, it's probably worth mentioning this in the documentation.

File: R/install_torsten.R [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+  require(cmdstanr)
+  if (check_toolchain) {
+    cmdstanr::check_cmdstan_toolchain(fix = FALSE, quiet = quiet)
+  }
+  if (is.null(dir)) {
+    dir <- file.path(dirname(cmdstanr::cmdstan_default_install_path()),
+                     ".torsten")
```
Comment (kyleam) [inline R/install_torsten.R:108]: I'd like to avoid littering `$HOME` with another app directory. I know cmdstanr does it, but I'm not convinced we should follow suit.

My preference would be to put these under `XDG_DATA_HOME` (e.g., `file.path(tools::R_user_dir("bbr.bayes"), "torsten")` or maybe just `file.path(tools::R_user_dir("torsten"))` ). `tools::R_user_dir()` isn't available until R 4.0, but that's okay for the purposes of bbr.bayes.

File: R/install_torsten.R [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    dir.create(dir, recursive = TRUE)
+  }
+  if (!is.null(version)) {
+    if (!is.null(release_url)) {
+      warning("version and release_url shouldn't both be specified!",
+              "\nrelease_url will be ignored.", call. = FALSE)
+    }
+
+    release_list <- get_torsten_release_list()
+    release <- release_list[grep(version, release_list)]
```
Comment (kyleam) [inline R/install_torsten.R:120]: `version` is user-supplied. They could pass something like "torsten" or even ".", and then this would give multiple values for `release`.

Instead we could 1) check that `version` is a scalar and 2) use `%in%` to verify that it's in the known set of releases.

File: R/install_torsten.R [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+                            ver, ".tar.gz")
+    }
+  }
+  message("* Installing Torsten from ", download_url)
+  dir_torsten <- file.path(dir, substr(basename(download_url), 1, nchar(basename(download_url)) - 7))
+  dir.create(dir_torsten, recursive = TRUE, showWarnings = FALSE)
+  dest_file <- file.path(dir_torsten, basename(download_url))
+  dir_cmdstan <- file.path(dir_torsten, "cmdstan")
+  ## Reset timeout for download. The 60 s default is not enough.
+  options(timeout = max(300, getOption("timeout")))
```
Comment (kyleam) [inline R/install_torsten.R:148]: This should restore `timeout` to its original value (e.g., using [withr::with\_options](https://withr.r-lib.org/reference/with_options.html)).

File: R/install_torsten.R [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+  #   cmdstanr::cmdstan_make_local(
+  #     dir = dir_cmdstan,
+  #     cpp_options = list(
+  #       CXX = "arch -arch arm64e clang++"
+  #     ),
+  #     append = TRUE
+  #   )
+  # }
+
+  message("* Building CmdStan binaries...")
```
Comment (kyleam) [inline R/install_torsten.R:187]: Although the current message isn't inaccurate (given Torsten builds on top of CmdStan), should this say "Torsten" instead?

File: tests/testthat/test-install-torsten.R [hunk 12]
```diff
@@ -0,0 +1,157 @@
+## Adapted from test-install.R from the cmdstanr package
+
+context("install_torsten")
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:3]: [context](https://testthat.r-lib.org/reference/context.html) has been deprecated; this can be safely dropped.

File: tests/testthat/test-install-torsten.R [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+torsten_test_tarball_url <- Sys.getenv("TORSTEN_TEST_TARBALL_URL")
+if (!nzchar(torsten_test_tarball_url)) {
+  torsten_test_tarball_url <- NULL
+}
+
+test_that("install_torsten() successfully installs torsten", {
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:10]: -   We've got a helper in bbr (`skip_if_over_rate_limit`) that skips the test if we've exhausted the GitHub API limits. bbr.bayes should be able to import that for use on its end (see `tests/testthat/setup-imports.R`).

-   It's not too uncommon to add a flag to skip network-dependent tests, but we don't have anything like that in bbr or bbr.bayes yet. We might want to eventually do that for these tests, but for now, let's at least mark the tests that actually do the installation as "long" tests by adding something like
    
    ```
    skip_long_tests("long-running install_torsten() tests")
    ```

File: tests/testthat/test-install-torsten.R [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  if (getRversion() < '3.5.0') {
+    dir <- tempdir()
+  } else {
+    dir <- tempdir(check = TRUE)
+  }
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:15]: -   I know this is `check = TRUE` part is coming from cmdstanr's tests, but we can drop it and assume that the per-session R temporary directory hasn't been deleted. Much of our code and tests use `tempfile`, and that calls `tempdir` with the default `check = FALSE`, so if something deleted the per-session R temporary directory, our tests aren't going to pass.

-   Ideally this would use a directory *under* `tempdir()` and clean it up afterward. You can use `local_test_dir()` for that, which is a bbr.bayes test helper that wraps `withr::local_tempdir()`.

File: tests/testthat/test-install-torsten.R [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+test_that("install_torsten() works with version and release_url", {
+  if (getRversion() < '3.5.0') {
+    dir <- tempdir()
+  } else {
+    dir <- tempdir(check = TRUE)
+  }
+
+  expect_message(
+    expect_output(
+      install_torsten(dir = dir, cores = 4,
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:82]: This and other spots should avoid assuming the test machine has a given number of `cores`. It can use the default, and the tester can adjust the `mc.cores` option as desired.

File: tests/testthat/test-install-torsten.R [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    "* Finished installing Torsten",
+    fixed = TRUE
+  )
+  expect_warning(
+    expect_message(
+      expect_output(
+        install_torsten(dir = dir, cores = 4,
+                        version = "0.89.1",
+                        # the URL is intentionally invalid to test that the version has higher priority
+                        release_url = "https://github.com/metrumresearchgroup/Torsten/archive/refs/tags/torsten_v0.89.1.tar.gz"),
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:96]: Did you forget to update this URL? (It's identical to the one above.)

File: tests/testthat/test-install-torsten.R [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    capture.output(install_torsten(dir = dir, quiet = TRUE),
+                   type = "message")
+  )
+
+##  expect_true(any(grepl(ver_msg, install_normal, fixed = TRUE)))
+  expect_true(any(grepl(download_msg, install_normal, fixed = TRUE)))
+
+##  expect_false(any(grepl(ver_msg, install_quiet, fixed = TRUE)))
+  expect_false(any(grepl(download_msg, install_quiet, fixed = TRUE)))
+})
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:137]: I'd vote to delete this test because it's not worth two installations to verify that we're plumbing `quiet` through correctly.

The worst case scenario of `quiet` not working properly is somehow triggering an `install_torsten` failure. That wouldn't be great, but 1) it'd be easy for the caller to work around by using the default `quiet` value (which we test) and 2) `install_torsten` is leaf functionality that wouldn't affect other things.

However, the code is simply passing `quiet` through to cmdstanr, so that sort of breakage is unlikely. The more likely thing is just that `quiet` doesn't have an affect (e.g, because we're not properly propagating it), and then the risk is just a noisier install.

(In addition, I'd like to avoid test code that matches messages from outside packages, which they are free to change without worrying about backward incompatibility.)

File: tests/testthat/test-install-torsten.R [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+#   }
+#
+#   expect_error({
+#     # Use an invalid proxy address to force a download failure
+#     withr::with_envvar(
+#       c("http_proxy"="invalid","https_proxy"="invalid"),
+#       install_torsten(dir = dir)
+#     )},
+#     "GitHub download of release list failed with error: cannot open URL 'https://api.github.com/repos/stan-dev/cmdstan/releases/latest'")
+# })
```
Comment (kyleam) [inline tests/testthat/test-install-torsten.R:156]: This commented out test can be deleted.

**Review by kyleam — CHANGES_REQUESTED** [review]
Thanks @billgillespie. Looks like a good start.

A few comments in addition to inline ones:

-   ****CI failure****: The reason for the [CI failure](https://github-drone.metrumrg.com/metrumresearchgroup/bbr.bayes/329/1/4) is a little convoluted and isn't directly related to the changes here. For the PR Drone run, the PR gets merged into the base, `main`. `main` had code changes that required changing `.drone.yml` (in particular, installing nmrec), but the *PR's* `.drone.yml` is in effect, so nmrec isn't available.
    
    So, we need `main`'s Drone configuration on this side. The simplest fix is just to update this branch so that it's based on the latest main (currently it's 40 commits behind). You can do that by rebasing on top of `origin/main` or merging `origin/main` into this PR, whichever you prefer.

-   ****stanc****: I don't think this is an issue to be resolved by `install_torsten`, but it's something I thought about while reviewing this. The CmdStan tar.gz is a custom release asset that bundles stanc:
    
    ```
    $ tar --list -f cmdstan-2.32.2.tar.gz | grep bin | grep stanc
    cmdstan-2.32.2/bin/windows-stanc
    cmdstan-2.32.2/bin/linux-stanc
    cmdstan-2.32.2/bin/mac-stanc
    ```
    
    The Torsten releases just have a source tar.gz, so during the installation process the *nightly* stanc is downloaded (can be overridden by setting `STANC3_VERSION` environment variable):
    
    ```
    $ pwd
    /data/home/kylem/.torsten/torsten_v0.90.0/cmdstan
    $ ./bin/stanc --version
    stanc3 2c6fa27 (Unix)
    ```
    
    I think that isn't ideal for at least two reasons: 1) it introduces an installation discrepancy regardless of the identical Torsten version being requested, potentially leading to different results and 2) the nightly stanc version may be incompatible with an old version of the toolchain being installed (Torsten/CmdStan/Stan).
    
    Perhaps Torsten should bundle a specific stanc version in the same way CmdStan does?

-   ****Inconsistent tag naming****: The latest two Torsten releases prefix the tag with `torsten_`, but the one before that doesn't. I think `install_torsten` should smooth over that discrepancy for the caller. For example, if they specify `v0.90.0` as the version, I think that should be expanded underneath to `torsten_v0.90.0`.
    
    I think it's safe to assume there should never be a conflicting release tag (e.g., both `torsten_v0.90.0` and `v0.90.0` won't exist), so we could just error if we somehow hit that case.

    *Update*: I missed that the current approach of using `grep` is already intended to address this (by allowing `torsten_v0.90.0` or any substring).  It just needs to be reworked a bit with a guard that ensures only one release is selected: https://github.com/metrumresearchgroup/bbr.bayes/pull/95#discussion_r1281038479

File: R/install_torsten.R (hunk 9 above)
Comment (kyleam) [reply R/install_torsten.R:120]: Update after chat with @billgillespie : part of the thinking of using `grep` was that it allows getting around the tag naming inconsistency (i.e. caller doesn't need to pass "torsten_").  So, sticking with `grep` makes sense; it'd just need to signal an error if more than one matching release is found.  Also, this should probably used `fixed = TRUE` so that `version` isn't treated as a regex.

File: R/install_torsten.R (hunk 1 above)
Comment (billgillespie) [reply R/install_torsten.R:12]: I moved `latest_torsten_release` and `get_torsten_release_list` to the end. They will remain unexported. 

Thanks for the tip on generating the documentation and `NAMESPACE`.

File: R/install_torsten.R (hunk 2 above)
Comment (billgillespie) [reply R/install_torsten.R:41]: I removed the apparently redundant comment hashes. Not sure why they were there.

File: R/install_torsten.R (hunk 3 above)
Comment (billgillespie) [reply R/install_torsten.R:55]: Done

File: R/install_torsten.R (hunk 4 above)
Comment (billgillespie) [reply R/install_torsten.R:56]: Do you mean that I should replace `tempfile` with `withr::local_tempfile`? If so, I did that. The doc for `withr::local_tempfile` is pretty terse. Am I right in thinking that it will delete `dest_file` immediately upon exiting the `latest_torsten_release` function?

Comment (kyleam) [reply R/install_torsten.R:56]: > Am I right in thinking that it will delete `dest_file` immediately upon exiting the `latest_torsten_release` function?

Correct.  It's functionally equivalent to

```
f <- tempfile(...)
on.exit(unlink(f), add = TRUE)
```

File: tests/testthat/test-install-torsten.R (hunk 15 above)
Comment (billgillespie) [reply tests/testthat/test-install-torsten.R:82]: How about the alternative of using `parallel::detectCores`, e.g., `install_torsten(dir = dir, cores = min(parallel::detectCores(), 4)` or `install_torsten(dir = dir, cores = min(parallel::detectCores(), getOption("mc.cores"))`? That way the test can take advantage of available cores even if the user has not set the `mc.cores` option.

File: tests/testthat/test-install-torsten.R (hunk 17 above)
Comment (billgillespie) [reply tests/testthat/test-install-torsten.R:137]: Works for me. For the moment I commented it out.

File: R/install_torsten.R (hunk 5 above)
Comment (billgillespie) [reply R/install_torsten.R:88]: `timeout` description added

File: R/install_torsten.R (hunk 7 above)
Comment (billgillespie) [reply R/install_torsten.R:100]: Added a brief statement in the description section of the documentation.

File: tests/testthat/test-install-torsten.R (hunk 15 above)
Comment (kyleam) [reply tests/testthat/test-install-torsten.R:82]: Hmm, I do see the appeal of letting developers get a speedup without manually configuring `mc.cores`.  While I don't think `mc.cores` is too surprising of a knob to tweak given the `install_cmdstan` and `install_torsten` default for cores is `getOption("mc.cores", 2)`, developers likely aren't going to go looking for it at all.

However, a complication here is choosing a method that keeps the tests well behaved in various scenarios, such as running `R CMD check` with `_R_CHECK_LIMIT_CORES_=TRUE` (as `devtools::test()` does by default), running in a container with a CPU quota, or running on a grid worker with the available CPUs set by scheduler (as is the case for test run by MPN build process).

Your suggestions would probably work well for most manual testing, aside from the edge case of `detectCores` returning `NA`.  Even for `cores = min(parallel::detectCores(), 4)` on machine with 4 cores or fewer, we're probably not worried about competing with too much else that's running on a manual tester's machine.

On the other hand, those suggestions wouldn't play nicely with some of the testing scenarios I mentioned above.  For example, under `_R_CHECK_LIMIT_CORES_=TRUE`, the tests would fail.

So, perhaps we can optionally use [parallelly::availableCores](https://parallelly.futureverse.org/reference/availableCoresa.html), which accounts for these things.  We could:

 * add `parallelly` the set of suggested packages in `DESCRIPTION`
 * add a test helper that uses `availableCores` if it's installed

The test helper could be something like this (untested):

```R
available_cores <- function() {
  if (!requireNamespace("parallelly", quietly = TRUE)) {
    return(getOption("mc.cores", 2))
  }
  return(unname(parallelly::availableCores(omit = 1)))
}
```

File: R/install_torsten.R (hunk 4 above)
Comment (billgillespie) [reply R/install_torsten.R:56]: Done

File: R/install_torsten.R (hunk 8 above)
Comment (billgillespie) [reply R/install_torsten.R:108]: I set the default to `file.path(tools::R_user_dir("torsten"))` and revised the description accordingly.

File: R/install_torsten.R (hunk 9 above)
Comment (billgillespie) [reply R/install_torsten.R:120]: Added a test for the case where multiple releases are matched by the `version` argument:
`    if(length(release) > 1){`
`      stop("The version argument, ", version, ", matches multiple Torsten version names", `
`           call. = FALSE)`
`    }`

File: R/install_torsten.R (hunk 11 above)
Comment (billgillespie) [reply R/install_torsten.R:187]: Changed to "Torsten".

File: R/install_torsten.R [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+  cmdstanr::rebuild_cmdstan(dir = dir_cmdstan, cores = cores,
+                            quiet = quiet, timeout = timeout)
+
+  set_cmdstan_path(path = file.path(dir_cmdstan))
+  message("Testing installation by attempting to compile an example model.\n")
```
[the remark this answers is not in the document]
Comment (billgillespie) [reply R/install_torsten.R:192]: Done

File: tests/testthat/test-install-torsten.R [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+        fixed = TRUE
+      ),
+      "* Finished installing Torsten",
+      fixed = TRUE
+    ),
+    "version and release_url shouldn't both be specified",
+    fixed = TRUE
+  )
+  expect_true(dir.exists(file.path(dir, "torsten_v0.89.1")))
+})
```
[the remark this answers is not in the document]
Comment (kyleam) [reply tests/testthat/test-install-torsten.R:107]: I'm not sure how well I conveyed the idea in prose, but, once your changes settle, I'd be happy to make the proposed changes on a branch on top of yours that you can merge into your branch.

[the remark this answers is not in the document]
Comment (billgillespie) [reply tests/testthat/test-install-torsten.R]: I pushed the changes made to date. There is still unfinished business...

File: R/install_torsten.R (hunk 10 above)
Comment (billgillespie) [reply R/install_torsten.R:122]: Done

File: R/install_torsten.R (hunk 6 above)
Comment (billgillespie) [reply R/install_torsten.R:69]: I implemented `overwrite`.

File: R/install_torsten.R [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+  message("* Installing Torsten from ", download_url)
+  dir_torsten <- file.path(dir, substr(basename(download_url), 1, nchar(basename(download_url)) - 7))
+  dir.create(dir_torsten, recursive = TRUE, showWarnings = FALSE)
+  dest_file <- file.path(dir_torsten, basename(download_url))
+  dir_cmdstan <- file.path(dir_torsten, "cmdstan")
+  ## Reset timeout for download. The 60 s default is not enough.
+  options(timeout = max(300, getOption("timeout")))
+  tar_downloaded <- try(suppressWarnings(utils::download.file(url = download_url,
+                                                              destfile = dest_file, quiet = quiet)),
+                        silent = TRUE)
```
[the remark this answers is not in the document]
Comment (billgillespie) [reply R/install_torsten.R:125]: Keeping `cmdstanr` behavior.

File: R/install_torsten.R [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+  message("* Unpacking archive...")
+  untar_rc <- utils::untar(
+    dest_file,
+    exdir = dir_torsten,
+    extras = "--strip-components 1"
+  )
+  if (untar_rc != 0) {
+    stop("Problem extracting tarball. Exited with return code: ", untar_rc, call. = FALSE)
+  }
+  file.remove(dest_file)
```
[the remark this answers is not in the document]
Comment (billgillespie) [reply R/install_torsten.R:148]: Done

File: tests/testthat/test-install-torsten.R (hunk 12 above)
Comment (billgillespie) [reply tests/testthat/test-install-torsten.R:3]: done

File: tests/testthat/test-install-torsten.R [hunk 23]
```diff
@@ -0,0 +1,157 @@
+## Adapted from test-install.R from the cmdstanr package
+
+context("install_torsten")
+
+torsten_test_tarball_url <- Sys.getenv("TORSTEN_TEST_TARBALL_URL")
+if (!nzchar(torsten_test_tarball_url)) {
+  torsten_test_tarball_url <- NULL
+}
```
[the remark this answers is not in the document]
Comment (billgillespie) [reply tests/testthat/test-install-torsten.R:8]: This came from `cmdstanr`. I think the idea is to allow the tester to use `TORSTEN_TEST_TARBALL_URL` to test specific versions that are not otherwise exercised by default in the tests. I just adapted it without thinking too hard about it.

[the remark this answers is not in the document]
Comment (kyleam) [reply tests/testthat/test-install-torsten.R:8]: Okay, fine with me if you want to keep it around.

**Comment by kyleam** [comment]
@billgillespie I've pushed some updates.

https://github.com/metrumresearchgroup/bbr.bayes/compare/49f87c9b675a0594d635887ca6c3fe4a7aa7f7b6...b67f16f166559bc7865f7fdc304b072bb4e91e65

Summary of more significant changes:

 * fix bug where `release_url` was unintentionally given precedence over `version` (fedcfdd)

   This was the reason for the [failing test](https://github-drone.metrumrg.com/metrumresearchgroup/bbr.bayes/337/1/4).

 * skip tests if over GitHub API limit (d1cdd2b)

 * additional test cases: default `dir` handling (07dc88d), `overwrite` handling (6e72c6b), "multiple matches" error (0b17a4c)

 * pull out `version` and `release_url` processing into a helper so that we can test the logic while limiting the tests to one full installation (04fcce9, da5418b)

Assuming that the tests pass on Drone, I think this should be in a good spot to merge.  Whenever you have a moment, please take a look over the changes, and let me know if they look fine to you.

**Review by kyleam — APPROVED** [review]

**Comment by kyleam** [comment]
I said:

> Assuming that the tests pass on Drone, I think this should be in a good spot to merge.

All green.

> Whenever you have a moment, please take a look over the changes, and let me know if they look fine to you.

Offline @billgillespie says "merge away".
